R/calculate_transition_disruption_metric.R

```diff
@@ -0,0 +1,121 @@
+
+  groups <- c(
+    "technology",
+    "ald_sector",
+    ...
+  )
+
+  data <- data %>%
+    # TODO: This function will only work if the allocation method is portfolio_weight
+    # ownership_weight outputs 0 for all carsten metric values, and thus we would be
+    # dividing by zero otherwise...
+    dplyr::filter(.data$allocation == "portfolio_weight") %>%
+    warn_if_has_zero_rows(
+      'Filtering for "portfolio_weight" allocation, outputs 0 rows'
+    )
+
+  if (nrow(data) == 0) {
+    return(empty_calculate_transition_disruption_metric())
+  }
+
+  technology_level_dy <- data %>%
+    dplyr::filter(.data$year %in% c(start_year, start_year + 5, start_year + 10)) %>%
+    dplyr::mutate(
+      time_step = dplyr::case_when(
+        .data$year == start_year ~ "start_year",
+        .data$year == start_year + 5 ~ "plus_five",
+        .data$year == start_year + 10 ~ "plus_ten"
+      )
+    ) %>%
+    dplyr::group_by(!!!rlang::syms(groups)) %>%
+    dplyr::select(
+      !!!rlang::syms(groups),
+      .data$time_step,
+      scen_alloc = "scen_alloc_wt_tech_prod",
+      plan_alloc = "plan_alloc_wt_tech_prod"
+    ) %>%
+    tidyr::pivot_wider(
+      names_from = .data$time_step,
+      values_from = c("scen_alloc", "plan_alloc")
+    ) %>%
+    dplyr::mutate(
+      .numerator = .data$scen_alloc_plus_ten - .data$plan_alloc_plus_five,
+      .denominator = .data$scen_alloc_plus_ten - .data$scen_alloc_start_year,
+      tdm_tech = max(0, .data$.numerator / .data$.denominator) * 2
+    ) %>%
+    dplyr::select(.data$tdm_tech, !!!rlang::syms(groups)) %>%
+    dplyr::ungroup()
+
+  data %>%
+    dplyr::filter(.data$year == start_year) %>%
+    dplyr::left_join(technology_level_dy) %>%
+    dplyr::group_by(!!!rlang::syms(groups)) %>%
+    dplyr::ungroup(.data$technology) %>%
+    dplyr::mutate(
+      tdm_sec = .data$plan_carsten * sum(.data$tdm_tech) / sum(.data$plan_carsten),
+      reference_year = start_year
+    ) %>%
+    dplyr::ungroup() %>%
+    dplyr::select(!!!rlang::syms(groups), .data$tdm_tech, .data$tdm_sec, .data$reference_year)
+}
+
+check_crucial_names <- function(data, expected_names) {
+  stopifnot(rlang::is_named(data))
+  stopifnot(is.character(expected_names))
+
+  ok <- all(unique(expected_names) %in% names(data))
+  if (!ok) {
+    abort_missing_names(sort(setdiff(expected_names, names(data))))
+  }
+
+  invisible(data)
+}
+
+abort_missing_names <- function(missing_names) {
+  nms <- glue::glue_collapse(missing_names, sep = ", ", last = ", and ")
+  abort(glue("Must have missing names:\n{nms}."), class = "missing_names")
+}
+
+warn_if_has_zero_rows <- function(data, message) {
+  if (nrow(data) == 0L) warn(message = message, class = "has_zero_rows")
+  invisible(data)
+}
+
+empty_calculate_transition_disruption_metric <- function() {
+  tibble::tibble(
+    technology = character(0),
+    ald_sector = character(0),
+    tdm_tech = numeric(0),
+    tdm_sec = numeric(0),
+    reference_year = integer(0),
```

Comment:
Note to self: It seems that these names could be reused, e.g. to write a test that checks the output has expected names.